ext/rugged/rugged_allocator.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) the Rugged contributors.  All rights reserved.
+ *
+ * This file is part of Rugged, distributed under the MIT license.
+ * For full terms see the included LICENSE file.
+ */
+
+#include "rugged.h"
+#include <git2/sys/alloc.h>
+
+static void *rugged_gmalloc(size_t n, const char *file, int line)
+{
+	return xmalloc(n);
+}
+
+static void *rugged_gcalloc(size_t nelem, size_t elsize, const char *file, int line)
+{
+	return xcalloc(nelem, elsize);
+}
+
+static char *rugged_gstrdup(const char *str, const char *file, int line)
+{
+	char *newstr;
+	size_t n;
+
+	n = strlen(str);
+	newstr = xmalloc(n+1);
+	memcpy(newstr, str, n);
+	newstr[n] = '\0';
+
+	return newstr;
```

Comment:
[Ruby implements a strdup](https://github.com/ruby/ruby/blob/e75fbfdacbf7d064dfcd908e222418a2da3869a5/util.c#L495-L505) so we can use that.

Reply:
I saw that, but it wouldn't find it when building, so I thought it was internal. Do we need to add an `#include`?

Reply:
Ya, it's public.  We just have to `#include <ruby/util.h>`

Reply:
Done via https://github.com/libgit2/rugged/pull/756/commits/f049f6d5f3fd12dc0d3a138239b308f7fc694aed.